File: web/selenium-python/utils/logger.py
import logging
import os

LOG_DIR = os.path.join(os.path.dirname(__file__), '..', 'logs')
os.makedirs(LOG_DIR, exist_ok=True)

logger = logging.getLogger('e2e')
logger.setLevel(logging.DEBUG)
_fmt = logging.Formatter('[%(asctime)s] [%(levelname)s] %(message)s')
_ch = logging.StreamHandler()
_ch.setFormatter(_fmt)
logger.addHandler(_ch)
_fh = logging.FileHandler(os.path.join(LOG_DIR, 'e2e.log'))
_fh.setFormatter(_fmt)
logger.addHandler(_fh)

File: web/selenium-typescript/utils/logger.ts
import * as fs from 'fs';
import * as path from 'path';

const LOG_DIR = path.join(__dirname, '../logs');
const LOG_FILE = path.join(LOG_DIR, 'e2e.log');
if (!fs.existsSync(LOG_DIR)) fs.mkdirSync(LOG_DIR, { recursive: true });

function write(level: string, message: string) {
  const line = `[${new Date().toISOString()}] [${level}] ${message}`;
  console.log(line);
  fs.appendFileSync(LOG_FILE, line + '\n');
}

export const logger = {
  info: (msg: string) => write('INFO', msg),
  warn: (msg: string) => write('WARN', msg),
  error: (msg: string) => write('ERROR', msg),
  debug: (msg: string) => write('DEBUG', msg),
};